src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -82,13 +83,21 @@ internal DocumentInfo(DocumentAttributes attributes, TextLoader? loader, IDocume
             string? filePath = null,
             bool isGenerated = false)
         {
-            return Create(id, name, folders, sourceCodeKind, loader, filePath, isGenerated, documentServiceProvider: null);
+            return Create(
+                id ?? throw new ArgumentNullException(nameof(id)),
+                name ?? throw new ArgumentNullException(nameof(name)),
```

Comment:
i have mixed feelings here.  One benefit is definitely that this is the public entrypoint, so we shoudl do arg validation here.  One downside is htat i still want the core constructors to do invariant checking (in case, for example, if Roslyn bypasses this entrypoint).

So i'm ok with what's going on here *if* the internal types/constructors still do some runtime validation.

Reply:
I'd agree with that if this was a more complex invariant to check - then we would want to have `Debug.Assert`. But we do have non-null invariants checked by the compiler via nullable annotations. I don't think adding non-null assertions is adding much value.

Reply:
Ok :)  hopefully nothing sneaks in.  I'd feel better if we used NRT in more of the code.  i.e. we won't catch a case where a non-NRT files calls directly into one of these non-checking constructors.  However, hopefully that's only a limited time issue as we spread NRT everywhere :)

Reply:
If nothing else, this pattern sets us up nicely for an automated refactoring to move the argument null checking to the language feature when it comes, and enforcing that all public methods are using that feature is probably what we'd ultimately do.